### ISSUE TITLE
Action runner loses pipe and causes shell script with redirections to hang

### ISSUE THREAD

**Issue by josefbacik** [actions/runner#2684]
**Describe the bug**
When running a script (https://github.com/btrfs/linux/actions/runs/5490218255/workflow) which runs fstests on a self-hosted VM, one of the tests will hang.  This test is doing something like this

_fail() {
   echo $1
   exit 1
}

od /dev/urandom | dd of=/some/file bs=1M count=10 > somefile.txt 2>&1 || _fail "dd failed"

The boxes executing this line get stuck here, if you look at the /proc/pid/fd listing you can see

[root@xfstests6 fstests]# ll /proc/33527/fd
total 0
lr-x------ 1 root root 64 Jul  7 21:13 0 -> 'pipe:[51951]'
l-wx------ 1 root root 64 Jul  7 21:13 1 -> /root/fstests/results/btrfs_block_group_tree/btrfs/016.full
l-wx------ 1 root root 64 Jul  7 21:13 10 -> /tmp/30763.out
lr-x------ 1 root root 64 Jul  7 21:13 106 -> 'pipe:[53318]'
l-wx------ 1 root root 64 Jul  7 21:13 11 -> /tmp/30763.out
l-wx------ 1 root root 64 Jul  7 21:13 112 -> 'pipe:[53319]'
l-wx------ 1 root root 64 Jul  7 21:13 2 -> /root/fstests/results/btrfs_block_group_tree/btrfs/016.full
lr-x------ 1 root root 64 Jul  7 21:13 255 -> /root/fstests/tests/btrfs/016

and the ps auxw show

root       30742  5.7  2.6 3707952 102100 ?      Sl   21:12   0:01 /root/actions-runner/bin/Runner.Worker spawnclient 106 112

so we're using those pipe's to capture the output of the job.

If I change this command to not do the redirects everything works fine.

Also I had a version of this runner that ssh'ed into the vm from a different machine to run this command, and this worked fine.  It only breaks when the runner is on the machine that is trying to run this testsuite.

**To Reproduce**
1. Start a local runner with fstests setup for btrfs (I can provide a series of steps to set this up if needed.)
2. Create a workflow that runs ./check btrfs/016
3. The task will hang

**Expected behavior**
The task should run just fine

## Runner Version and Platform
actions-runner-linux-x64-2.305.0

OS of the machine running the runner? OSX/Windows/Linux/...
Fedora 38

## What's not working?
The workflow is hanging when the runner is local to the machine running the testsuite.

## Job Log Output
You can see my attempt here https://github.com/btrfs/linux/actions/runs/5490218255.

## Runner and Worker's Diagnostic Logs
The logs don't show anything helpful, but I can provide them if needed.

**Comment by josefbacik** [comment]
I narrowed it down, if you create a simple worker and just do exactly that line it hangs, so add a file called reproducer.sh, add this to it

```
#!/bin/bash

_fail() {
        echo $*
        exit 1
}

od /dev/urandom | dd iflag=fullblock of=/tmp bs=1M count=10 >> /log.txt 2>&1 || _fail "failed to dd"
```

and then setup a simple worker to run that and it hangs.

**Comment by josefbacik** [comment]
Turned out to be a bug in od, it'll happily keep writing to a pipe that's closed and never exits out.  I'm still confused why this particular setup fails but the script outside of github actions works fine, but that's a problem for somebody else to figure out.

**Comment by josefbacik** [comment]
Turns out I'm dumb.

Things like

command1 | command2

work because of SIGPIPE, one side or the other stops, the other one continues to operate on the pipe, it gets SIGPIPE and it exits, no muss no fuss.  In the reproducer

od /dev/urandom | dd of=somefile bs=1M count=1

od will run forever, because /dev/urandom is a special device that will never EOF, so the above case relies on dd exiting and then od exiting because it gets SIGPIPE.

However there are two things preventing this from happening.

1. systemd.  When you run the action runner as a service in fedora it installs a systemd file to launch it.  By default, systemd masks off SIGPIPE, and children from fork() inherit all signal masking, so any process launched from systemd will inherit ignoring SIGPIPE.  This can be fixed with adding `IgnoreSIGPIPE=no` to the systemd file.  However this isn't enough.
2. node.js.  The action runner is launched via node.js, and according to this documentation (https://nodejs.org/api/process.html#signal-events) SIGPIPE is masked off.  I confirmed this by checking the script that systemd launches doesn't have SIGPIPE masked when I have `IgnoreSIGPIPE=no` set, but the worker itself which is launched via node.js does have SIGPIPE masked.  This is where me being a kernel engineer fails me, also it being a weekend and me being lazy, I couldn't figure out how to make node.js to the equivalent of `signal(SIGPIPE, SIG_DFL)`, it only looks like you can attach your own signal handler.  This may be good enough since the children don't inherit the signal handlers, so at the very least it wouldn't be masked off, but I'm not the guy to make that change.

Since xfstests is just a series of shell scripts, doing something like `trap - SIGPIPE` in the main script doesn't do anything, bash won't reset ignored signals if it's already ignored before running the script.  I've had to work around this issue by writing a small C wrapper to launch xfstests from GitHub Actions

```
#include <signal.h>
#include <unistd.h>
#include <stdio.h>
#include <sys/wait.h>

int main(int argc, char **argv)
{
	pid_t child;

	if (argc == 1) {
		fprintf(stderr, "Must specify a command to run");
		return 1;
	}

	child = fork();
	if (child < 0) {
		fprintf(stderr, "Couldn't fork");
		return 1;
	}

	if (child) {
		int wstatus;

		wait(&wstatus);
		if (WIFEXITED(wstatus))
			return WEXITSTATUS(wstatus);
		if (WIFSIGNALED(wstatus))
			raise(WTERMSIG(wstatus));
		return 1;
	}

	signal(SIGPIPE, SIG_DFL);
	return execvp(argv[1], argv+1);
}
```

With this I can run

```
./fix-signal ./check -g auto
```

from my GitHub actions script and everything works fine.

I had looked at changing Runner.Sdk/ProcessInvoker.cs ExecuteAsync() to do the signal(SIGPIPE, SIG_DFL) thing before running the command, that would likely be the cleanest solution, but it appears C# doesn't know what SIGPIPE is, and googling around the PosixSignal stuff that got added recently didn't include SIGPIPE, so I'm not sure how to fix this in a clean way as I'm not a C# developer either.

It would be good to have this fixed, I'm good for now but this would be kind of annoying for any other person who would run into this.

**Comment by nikola-jokic** [comment]
Hey @josefbacik,

Thank you for explaining everything in the comments! However, I don't think the runner should be changed here. The SIGPIPE does not work the same way on Windows as it would work on Unix-like operating systems. The runner is just an agent trying to execute whatever you tell it to execute. It also supports many platforms and can be started in many ways. So trying to infer how it is started, what signals are available on multiple platforms can increase complexity (and likely decrease performance) so much that I don't think it is feasible.

**Comment by josefbacik** [comment]
I agree it would be annoying to deal with all of the signals, but really SIGPIPE is the only one you'd have to worry about in real life.  99% of the time most test suites aren't going to use pipes, but pipes aren't that rare, a lot of scripts use them.  Even for btrfs-progs which is purely userspace we have pipes in some of the scripts for generating corrupt file systems to test btrfs check against.

Additionally you already have platform specific code throughout this project.  The cleanest solution would be to take what I've done and adapt it into ExecuteAsync() after the fork().  That would keep the blast radius down and wouldn't mar up the code too much.

But I don't see how you could do this with C#, and adding in some random C dependency to accomplish it wouldn't be fun for portability.  If you guys don't want to fix this that's fine, I think simply having it documented here (and maybe a note in the documentation) is enough.  I very much doubt there's a lot of kernel engineers as insane as me trying to get GitHub actions to do this, and if they are it'll be because I told them to try it and have already warned them about this.

**Comment by nikola-jokic** [comment]
Thanks for the answer :relaxed: I will raise this issue with the team and get back to you with the decision :relaxed:

**Comment by socketbox** [comment]
@josefbacik Hats off to you for the troubleshooting and sleuthing you did.

@nikola-jokic the conceit that you've provided for the user is that they're running in an actual shell, with the full functionality of that environment. 

If I want to be explicit about it, I can, when defining a step, indicate that the code in `run` should be executed in `shell: bash` (https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsshell). 

If the shell that the code is executed within is somehow hobbled by an implementation decision involving nodejs, then please document that fact -- it would have saved me some time.
